src/util/psbt/psbtmap/psbtglobal.rs

```diff
@@ -0,0 +1,137 @@
+        for txin in &tx.input {
+            if !txin.script_sig.is_empty() {
+                return Err(util::Error::from(Error::UnsignedTxHasScriptSigs));
+            }
+
+            if !txin.witness.is_empty() {
+                return Err(util::Error::from(Error::UnsignedTxHasScriptWitnesses));
+            }
+        }
+
+        Ok(PsbtGlobal {
+            unsigned_tx: tx,
+            unknown: Default::default(),
+        })
+    }
+}
+
+impl PsbtMap for PsbtGlobal {
+    fn insert_pair(&mut self, pair: PsbtRawKVPair) -> Result<(), util::Error> {
+        let PsbtRawKVPair {
+            key: raw_key,
+            value: raw_value,
+        } = pair;
+
+        match raw_key.type_value {
+            0u8 => {
+                return Err(util::Error::from(Error::DuplicateKey(raw_key)));
+            }
+            _ => {
+                if self.unknown.contains_key(&raw_key) {
+                    return Err(util::Error::from(Error::DuplicateKey(raw_key)));
+                } else {
+                    self.unknown.insert(raw_key, raw_value);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_pairs(&self) -> Result<Vec<PsbtRawKVPair>, util::Error> {
+        let mut rv: Vec<PsbtRawKVPair> = Default::default();
+
+        rv.push(PsbtRawKVPair {
+            key: PsbtRawKey {
+                type_value: 0u8,
+                key: vec![],
+            },
+            value: self.unsigned_tx.psbt_serialize()?,
+        });
+
+        for (key, value) in self.unknown.iter() {
+            rv.push(PsbtRawKVPair {
+                key: key.clone(),
+                value: value.clone(),
+            });
+        }
+
+        Ok(rv)
+    }
+
+    fn merge(&mut self, other: Self) -> Result<(), util::Error> {
+        if self.unsigned_tx != other.unsigned_tx {
+            return Err(util::Error::ParseFailed);
+        }
+
+        self.unknown.extend(other.unknown);
+        Ok(())
+    }
+}
+
+impl From<Transaction> for Result<PsbtGlobal, util::Error> {
+    fn from(tx: Transaction) -> Self {
+        PsbtGlobal::from_unsigned_tx(tx)
+    }
+}
+
+impl_psbtmap_consensus_encoding!(PsbtGlobal);
+
+impl<D: SimpleDecoder<Error = util::Error>> ConsensusDecodable<D> for PsbtGlobal {
+    fn consensus_decode(d: &mut D) -> Result<Self, D::Error> {
+        let mut pairs: Vec<PsbtRawKVPair> = Default::default();
+
+        while let Ok(pair) = ConsensusDecodable::consensus_decode(d) {
+            pairs.push(pair);
+        }
+
+        let (tx_raw_kv_pair, unknown_raw_kv_pairs): (
+            Vec<PsbtRawKVPair>,
+            Vec<PsbtRawKVPair>,
+        ) = pairs
+            .into_iter()
+            .partition(|ref raw_kv_pair| raw_kv_pair.key.type_value == 0u8);
+
+        match tx_raw_kv_pair.len() {
+            0 => Err(D::Error::from(Error::MustHaveUnsignedTx)),
+            1 => {
+                let tx: Transaction = deserialize(&tx_raw_kv_pair[0].value[..])?;
```

Comment:
You need to check at some point that `raw_kv_pair.key.key.is_empty()`, because as written this will deserialize any key with the type_value 0x00, no matter its length or contents, which the BIP forbids.

Also, `Transaction::deserialize` will ignore trailing bytes, which you should flag an error on (and also flag an error if it's signed, has a witness array, etc). Annoyingly to catch trailing bytes you need to copy the `RawDecoder` logic out of `network::serialize::deserialize`, unwrap the rawdecoder when its done, and check that the underlying cursor is spent.

Reply:
Hmmm... Anything else that will ignore trailing bytes? I'd be happy to check for all of them

Reply:
Yes, the deserializer will ignore trailing bytes on all Bitcoin objects because they all self-describe their sizes, and the deserializer is meant to work with streams where "no more bytes" is not always well-defined.

Reply:
Hmmm... I think that makes sense for `ConsensusDecodable`, but perhaps `deserialize` should error if the cursor is not consumed as the caller is explicitly saying that the bytes provided should `deserialize` to a certain type?